### PR TITLE
Parse & cache queries

### LINKED ISSUE #100: finish experimental response post-processing
PR #86 reduces the performance when processing queries. In a benchmark, we saw that the breaking point went from 15k rps to 3k rps with that PR.

For now, #99 feature gates the response post processing, but this is a temporary solution, we need a proper fix soon.

For more context about the performance issue, here's a flamegraph of the `format_response` function:
Screenshot from 2021-11-08 17-04-05

That function takes about 3% of the total processing time, of which a large part happens in parsing the query and extracting selection sets (about 2.7% of the total processing time). The call to `apply_selection_set` that actually modifies the response takes about 0.22%, which is fine for now.

We might want to cache the result of the query parsing, but the selection set structure comes from apollo-rs, and contains references to the input data.
So caching that would require storing a copy of the query, and the apollo-rs `Document` that refers to it, and hope that we don't get too many issues with self referential structures.
Another path would be to have a completely owned structure to represent selection sets. The structure would not be too complex for now, but I worry this will collide with future work about query planning and execution.

Any thoughts on this?

### PR DESCRIPTION
**Pull request by cecton**
## Changes
- Removal of feature flag `post-processing`
- Introduction of new object `QueryCache` in core crate, shamelessly copy-pasting the logic from `CachingQueryPlanner` (will be refactored in a future PR)
- Move `Query` to its own file (`query.rs`)
- Add CLI argument `--query-cache-limit`
- Query parsing in parallel to plan execution (because it's in a `spawn_blocking`)
- Regrets (refactor in a future PR)

## Task list from #106 
- [x] fragments can be defined in the supergraph (this is the most common case) so fragments definition should be obtained from there too, not only in the query
- [x] we should check the operation name, to apply the correct selection set (a query can contain multiple operations)

Superseeds #106 
Closes #100 
Related to #166

### CODE REVIEW

**Comment by cecton** [comment]
>  we should check the operation name, to apply the correct selection set (a query can contain multiple operations)

@Geal I have a question on this one. Can you give me a scenario where an incorrect selection set is applied??

I'm asking because the query defines all its operations so the selection set comes from the query itself, I don't see how it can be the wrong one?

**Comment by Geal** [comment]
there can be multiple named queries in one request, and the operation field decides which one is used. I think the query planner is already validating this though, so maybe we can avoid checking it again?

File: apollo-router-core/src/query_planner/mod.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-mod caching_query_planner;
-mod model;
-mod router_bridge_query_planner;
+pub(crate) mod caching_query_planner;
+pub(crate) mod model;
+pub(crate) mod router_bridge_query_planner;
```
Comment (garypen) [inline apollo-router-core/src/query_planner/mod.rs:3]: I think only the model needs to be crate public.

Comment (garypen) [inline apollo-router-core/src/query_planner/mod.rs:3]: I think only the model needs to be crate public.

**Review by garypen — CHANGES_REQUESTED** [review]
Looks good. One small change I think.

I had intended to polish up the caching stuff in a later PR to make it a reusable component. I think it can be generic over keys/values and should be fairly simple to achieve.

I note in your PR comment that you talk about re-factoring this later, so I think it's fine to duplicate in this PR and then we can re-factor the caching component later.

File: apollo-router-core/src/query_cache.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+use crate::prelude::graphql::*;
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::broadcast;
+
+/// A cache for parsed GraphQL queries.
+#[derive(Debug)]
+pub struct QueryCache {
```
Comment (Geal) [inline apollo-router-core/src/query_cache.rs:10]: considering how tricky that code can be, we should refactor that in a separate cache type

Comment (Geal) [reply apollo-router-core/src/query_cache.rs:10]: to have one canonical cache implementation

File: apollo-router-core/src/query.rs [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+            _ => {
+                failfast_debug!("Invalid type for data in response.");
+            }
+        }
+    }
+
+    pub fn parse(
+        query: impl Into<String>,
+        schema: &Schema,
+    ) -> tokio::task::JoinHandle<Option<Self>> {
```
Comment (Geal) [inline apollo-router-core/src/query.rs:51]: I'm not sure about returning a join handle here (there's nothing that warrants access to the join handle at https://github.com/apollographql/router/pull/205/files#diff-830a30ed2f89feb3a5d8d08941aa96390c492fcceccc7d8ec4564ae800a712a6R74 ). Could we convert that to a more abstract future and hide the task spawning here?

I think putting the parsing in a blocking task should even be one level above, in the QueryCache code

File: apollo-router-core/src/query_cache.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    cached: Mutex<LruCache<String, Option<Arc<Query>>>>,
+    #[allow(clippy::type_complexity)]
+    wait_map: Mutex<HashMap<String, broadcast::Sender<(String, Option<Arc<Query>>)>>>,
+    cache_limit: usize,
+    schema: Arc<Schema>,
+}
+
+impl QueryCache {
```
Comment (BrynCooke) [inline apollo-router-core/src/query_cache.rs:18]: Can we have some basic unit tests for this?

**Review by o0Ignition0o — COMMENTED** [review]
A good first step!

+1 for a more generic cache (with only one implementation)

I think we could also work on our router_factory builder to decrease the number of parameters create and recreate need, maybe as a followup ?

File: apollo-router-core/src/query_cache.rs (hunk 2 above)
Comment (cecton) [reply apollo-router-core/src/query_cache.rs:10]: Yes I agree. As I said I plan to do this in a later PR. Unless you think there is an argument for doing this here I will consider this out-of-scope for now.

File: apollo-router-core/src/query_cache.rs (hunk 4 above)
Comment (cecton) [reply apollo-router-core/src/query_cache.rs:18]: I propose we delay that for the PR that will refactor the caching thing as this should be done only once.

**Comment by cecton** [comment]
> +1 for a more generic cache (with only one implementation)

Yes I agree. As I said I plan to do this in a later PR. Unless you think there is an argument for doing this here I will consider this out-of-scope for now.

File: apollo-router-core/src/query_planner/mod.rs (hunk 1 above)
Comment (cecton) [reply apollo-router-core/src/query_planner/mod.rs]: Fixed in 33e03a6

Comment (cecton) [reply apollo-router-core/src/query_planner/mod.rs]: Fixed in 33e03a6

File: apollo-router-core/src/query.rs (hunk 3 above)
Comment (cecton) [reply apollo-router-core/src/query.rs]: Fair point.

Fixed in 44fdb3cfefbc2d279181e28b1e0b07a1466c57f0

**Review by o0Ignition0o — APPROVED** [review]
LGTM, once we have follow up issues on having one generic cache implementation with tests, and possibly on the router_factory refactoring (router_refactoring? 😁 )

**Comment by cecton** [comment]
Routerifactorio

**Review by garypen — APPROVED** [review]

File: apollo-router-core/src/query_cache.rs (hunk 2 above)
Comment (Geal) [reply apollo-router-core/src/query_cache.rs:10]: yep, it can be done in another PR

File: apollo-router-core/src/query.rs (hunk 3 above)
Comment (Geal) [reply apollo-router-core/src/query.rs]: nice, thanks

File: apollo-router-core/src/query_cache.rs (hunk 2 above)
Comment (BrynCooke) [reply apollo-router-core/src/query_cache.rs:10]: It's also worth checking that a loading cache implementation does not exist already. This one looks promising, although I can't vouch for the internals.

https://github.com/zerotwo-bot/cache-loader-async-rs

**Comment by BrynCooke** [comment]
This got squashed in another comment, but the question still stands. Did we check for an off the shelf cache implementation? If not why not?

**Comment by Geal** [comment]
We went with lru as it's commonly used. I proposed https://github.com/al8n/stretto but it's a bit young

**Comment by garypen** [comment]
> This got squashed in another comment, but the question still stands. Did we check for an off the shelf cache implementation? If not why not?

We are using an off the shelf cache implementation, lru. We've written a small amount of code around it to optimise for our particular use case.

**Comment by BrynCooke** [comment]
@cecton Has said this is going to be revisited, so I won't stand in the way of merging. Apparently it will get unit tested later.

I understand that this is a small wrapper around the commonly used LRUCache, But the fact remains that it is adding a bunch of async code, which honestly looks correct, but might not be.

I think the think that surprises me is that these isn't a good high quality implementation that does what we want in the Rust ecosystem already, it's such a common thing to want to do.

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**